### ISSUE TITLE
Quadratic performance on left-associated binds

### ISSUE THREAD

**Issue by tomjaguarpaw** [haskell-streaming/streaming#109]
# The problem

Consider walking over a tree using a streaming library to do something at every leaf:

```haskell
data Tree = Branch Tree Tree | Leaf Int

walkTreeTransformer :: (Monad (m IO), MonadTrans m)
                    => (Int -> IO ()) -> Tree -> m IO ()
walkTreeTransformer doSomething = loop
  where loop = \case
          Leaf i -> lift (doSomething i)
          Branch t1 t2 -> do
            loop t1
            loop t2
```

On left-skewed trees, i.e. trees that will end up generating left-associated binds, for example those generated by `leftSkewed`

```haskell
-- Fully evaluated left-skewed tree
leftSkewed :: Int -> Tree
leftSkewed 0 = Leaf 0
leftSkewed n = (Branch $! leftSkewed (n - 1)) (Leaf n)
```

the performance of your streaming library is quadratic. In fact performance is quadratic under

* streaming
* pipes
* conduit

(The only streaming library I know of whose performance is not quadratic is streamly.)

The plot below demonstrates the claim. The code to generate the plot is available at https://github.com/tomjaguarpaw/streaming-benchmark. The compiler was GHC 8.10.7.

![image](https://user-images.githubusercontent.com/1951567/135711881-d7a1320f-ac2f-4d98-8a4c-c4e7f8bc447d.png)

# The solution

Firstly, before talking about a solution, do you actually consider this a bug? I do, and I think the risk of hitting unexpected quadratic performance makes this library unsuitable for production use as it is. On the other hand maybe you have a good reason that I shouldn't think that. If so I would be interested to hear it.

If you *do* think this is a bug then let's think about what can be done. I know of two straightforward options:

1. Add [an explicit constructor for binds](https://github.com/tomjaguarpaw/streaming-benchmark/blob/master/app/Streaming/Better.hs#L9-L12) and then [reassociate binds during elimination](https://github.com/tomjaguarpaw/streaming-benchmark/blob/master/app/Streaming/Better.hs#L49-L62)
2. Use [`Codensity`](https://github.com/tomjaguarpaw/streaming-benchmark/blob/master/app/Streaming/Codensity.hs#L9) (streamly essentially uses [handwritten `Codensity`](https://hackage.haskell.org/package/streamly-0.8.0/docs/src/Streamly.Internal.Data.Stream.StreamK.Type.html#Stream).)

In my benchmarks explicit bind comes out slightly ahead.

Perhaps there is another option more suitable for your library.

I welcome your thoughts.

**Comment by harendra-kumar** [comment]
This paper https://rubenpieters.github.io/assets/papers/JFP20-pipes.pdf presents similar results, it has some more benchmarks as well. @tomjaguarpaw its nice to see that you have presented a solution as well. Looks neat.

**Comment by pranaysashank** [comment]
@tomjaguarpaw Here's another data point. While working on streamly, I experimented with changing streaming' s representation to be more like streamly's (and vector's) internal representation which can be found at [streaming-fusion](https://github.com/pranaysashank/streaming-fusion). 
Here is that library added to the benchmark:

![2021-10-03-131812_2068x1240_scrot](https://user-images.githubusercontent.com/15224715/135745011-3e1f40ee-d435-4ec8-b99f-68a1c7aead4c.png)

**Comment by treeowl** [comment]
@pranaysashank, I find the colors on that graph really hard to work out. Could you rearrange the key from top to bottom or something?

**Comment by pranaysashank** [comment]
@treeowl I don't know enough about gnuplot to modify the generated image. So I just updated the comment with the keys top to bottom

**Comment by tomjaguarpaw** [comment]
[N.B. conduit doesn't have quadratic performance. I made a mistake when setting up the benchmarks: https://github.com/tomjaguarpaw/streaming-benchmark/commit/2a965a790c23c759c70ac580b25c59fd22ae922c]

**Comment by tomjaguarpaw** [comment]
> While working on streamly, I experimented with changing streaming' s representation to be more like streamly's

Yup, makes sense that the performance is similar to streaming when I wrapped it with `Codensity`. I find it interesting that streaming with explicit binds is faster still!

**Comment by pranaysashank** [comment]
The Bind implementation is nice, I changed the MonadTrans instance of `streaming-fusion` to be more like the streaming bind implementation and looks like it improves the performance quite a bit!!

![image](https://user-images.githubusercontent.com/15224715/135748736-d368f882-70a5-40ad-9317-f8cf4cd8adaa.png)

**Comment by tomjaguarpaw** [comment]
@pranaysashank Wow! That's impressive.
